### ISSUE TITLE
How to change confidence threshold in model.train?

### ISSUE THREAD

**Issue by Nadayz** [ultralytics/ultralytics#11707]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, 

I want to change the confidence threshold in train mode; how can I do that?

model = YOLO("yolov8m.pt")
results = model.train(data="data.yaml",optimizer="Adam")

**Comment by jdiaz97** [comment]
I don't think that's a parameter https://docs.ultralytics.com/modes/train/#train-settings

**Comment by glenn-jocher** [comment]
Hello,

You're correct; the confidence threshold is not directly adjustable during the training phase using `model.train()`. The confidence threshold is typically used during inference to filter out detections based on their confidence scores. When training, the model learns to predict bounding boxes and class confidences, and the actual threshold to use can be set later during validation or prediction.

If you have further questions or need assistance with anything else, feel free to ask! Happy coding! 🚀

**Comment by glenn-jocher** [comment]
@jihunddok hello! Thank you for providing the clear summary of your scenario. It seems like the challenge you're facing is mainly about handling overlapping masks and prioritizing certain detections when multiple models predict different objects in similar areas.

One effective approach could be to implement a post-processing step where you can merge or prioritize overlapping masks based on certain criteria. Here’s a basic strategy:

1. **Intersection Over Union (IoU)** - Calculate the IoU for overlapping masks. If IoU exceeds a threshold (e.g., 0.5), you may consider them as overlapping.

2. **Confidence Score Priority** - In cases of overlap, keep the mask with the higher confidence score and discard the lower one.

3. **Class Priority List** - If you know certain objects (like 'pen' over 'pencil') are more likely or important, you can create a priority list. Use this list to determine which mask to keep when overlaps occur.

Here’s a simple pseudocode implementation:
```python
def process_predictions(predictions, confidence_threshold=0.5, iou_threshold=0.5):
    # Assuming predictions is a list of tuples (mask, confidence, class_id)
    # Sort predictions by confidence score
    predictions.sort(key=lambda x: x[1], reverse=True)

    final_masks = []
    for current_mask, current_conf, current_class in predictions:
        keep = True
        for final_mask, _, _ in final_masks:
            iou = calculate_iou(current_mask, final_mask)
            if iou > iou_threshold:
                keep = False
                break
        if keep:
            final_masks.append((current_mask, current_conf, current_class))

    return final_masks

# Utility function to calculate IoU
def calculate_iou(mask1, mask2):
    # Implement IoU calculation between two masks
    pass
```

This method does not require enhancing model performance but helps in intelligently managing the outputs from multiple models. Use OpenCV to draw masks from `final_masks` which should now have reduced overlaps and prioritized detections. If modifying overlapping areas based on model output is feasible, adjusting the `iou_threshold` and `confidence_threshold` in the process_predictions function could grant better control over the outcomes.

Please test and adapt the code as necessary for your specific application context. Hope this helps! Let me know if you have further questions or need more specific examples. Happy coding! 🚀

**Comment by Nadayz** [comment]
> Hello,
> 
> You're correct; the confidence threshold is not directly adjustable during the training phase using `model.train()`. The confidence threshold is typically used during inference to filter out detections based on their confidence scores. When training, the model learns to predict bounding boxes and class confidences, and the actual threshold to use can be set later during validation or prediction.
> 
> If you have further questions or need assistance with anything else, feel free to ask! Happy coding! 🚀

Thanks for your answer.
OK. when I try to train the model on my data. the precision is good, but Recall is very low.
How can I fix this problem?

[results.csv](https://github.com/ultralytics/ultralytics/files/15247693/results.csv)
